### ISSUE TITLE
Android performance: CPU affinity

### ISSUE THREAD

**Issue by alexcohn** [tensorflow/tensorflow#29910]
**System information**
- Mobile device: **Xiaomi A2 (8 core)**
- TensorFlow installed from (source or binary): **source**
- TensorFlow version (use command below): **master** at [6cf83ea] (2019/06/17)

**Describe the current behavior**

The `benchmark_model` that we run on the device performs very differently when we choose CPU affinity. E.g. average inference time (32 bit, CPU only) for our custom .tflite model is 172ms for taskset=ff, 143ms for taskset=f0, and 281ms taskset=0f. We used num_threads=8, i.e. much more threads that we have cores available. When we increase num_threads=12, inference time further decreases. The change of min times is even more impressive: 125ms for ff, 114ms for f0, and 211ms for 0f. And 105ms for f0 with 12 threads, i.e. 3 threads per CPU. Further increasing num_threads does not deliver visible improvements.

Note that we have no control on the actual number of threads used by interpreter->Invoke(), and their CPU affinity. 

**Describe the expected behavior**

I would expect the _interpreter_ to choose the optimal threading model. The [document](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/tools/benchmark#reducing-variance-between-runs-on-android) reads:

> Reducing variance between runs on Android.
> When running benchmarks on these phones there can be significant variance between different runs of the benchmark.

While benchmarks are nice, our real necessity is to run TFLite optimally in our app. But we cannot control thread CPU affinity or the _interpreter_?

But even if we could setup the _interpreter_ threading configuration beyond the generic _"Set the number of threads available to the interpreter"_,  across multiple devices, choosing the optimal taskset is beyond the capabilities of most development teams. This cannot be done by analyzing the cpuinfo: on our A2 development phone, all 8 cores are declared [almost](https://github.com/tensorflow/tensorflow/issues/29910#issuecomment-503589179) equal:

```
Processor       : AArch64 Processor rev 4 (aarch64)
processor       : 0
BogoMIPS        : 38.40
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32
CPU implementer : 0x51
CPU architecture: 8
CPU variant     : 0xa
CPU part        : 0x801
CPU revision    : 4

processor       : 1
BogoMIPS        : 38.40
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32
CPU implementer : 0x51
CPU architecture: 8
CPU variant     : 0xa
CPU part        : 0x801
CPU revision    : 4

processor       : 2
BogoMIPS        : 38.40
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32
CPU implementer : 0x51
CPU architecture: 8
CPU variant     : 0xa
CPU part        : 0x801
CPU revision    : 4

processor       : 3
BogoMIPS        : 38.40
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32
CPU implementer : 0x51
CPU architecture: 8
CPU variant     : 0xa
CPU part        : 0x801
CPU revision    : 4

processor       : 4
BogoMIPS        : 38.40
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32
CPU implementer : 0x51
CPU architecture: 8
CPU variant     : 0xa
CPU part        : 0x800
CPU revision    : 2

processor       : 5
BogoMIPS        : 38.40
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32
CPU implementer : 0x51
CPU architecture: 8
CPU variant     : 0xa
CPU part        : 0x800
CPU revision    : 2

processor       : 6
BogoMIPS        : 38.40
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32
CPU implementer : 0x51
CPU architecture: 8
CPU variant     : 0xa
CPU part        : 0x800
CPU revision    : 2

processor       : 7
BogoMIPS        : 38.40
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32
CPU implementer : 0x51
CPU architecture: 8
CPU variant     : 0xa
CPU part        : 0x800
CPU revision    : 2

Hardware        : Qualcomm Technologies, Inc SDM660

```
Maybe I am missing something, but I see here no clue that the performance of **f0** will be so much different from **0f**.

The dynamic optimization should be performed once in a while, because the load on the device may change, either in the same process, or because of other processes/apps running along our app.

It could be useful to persist these findings, so that next time interpreter starts, it could have a reasonable starting point. I am not sure if this info is relevant for specific .tflite model, or for any model. In our experiments, the we only used similar FCNN networks, and their performance was effected by taskset just the same.

**Comment by freedomtan** [comment]
If you check `cpuinfo` carefully, you can find that processors 0-3 and processor 4-7 are different.
For 0-3,
```
CPU part        : 0x801
CPU revision    : 4
```
For 4-7,
```
CPU part        : 0x800
CPU revision    : 2
```

Modern cell phone SoCs tend to have 2 or more clusters. All processors use the same ISA, but different cluster may have different properties, e.g., one cluster for power-efficiency and the other for computation performance. This is what called [big.LITTLE](https://www.arm.com/why-arm/technologies/big-little) by ARM.

**Comment by alexcohn** [comment]
Ok, so without running the benchmark on a specific device, we can know how to choose the performant CPU mask. But what is the recommended way to apply this when it happens in our Android app?

**Comment by freedomtan** [comment]
Nope, not all devices are created equal. There can be more than 2 clusters and there are different kinds of big and LITTLE cores. There could be different scheduling and cpufreq policies. It's complicated. I guess there is no general recommended way to choose number of threads that can be applied to all phones.

**Comment by alexcohn** [comment]
I guess I didn't ask the right question. Let me try again. Even if I know the optimal settings for this specific device, how can I regulate CPU affinity of inference threads?

**Comment by freedomtan** [comment]
in Java? as far as I can remember, there is not such API. In C, the `sched_setaffinity(2)` system call should work.

**Comment by alexcohn** [comment]
So, essentially you suggest to run tflite in a dedicated process. This may be a very cool way to set it up, but IPC overhead may outweigh performance gains.

**Comment by jdduke** [comment]
One thing to note is that is that our legacy, multi-threaded floating point kernels are notoriously sensitive to unpinned execution, and generally work poorly with Android's thread scheduler.

We've been working on a [new gemm backend](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/BUILD#L193) which largely resolves this, and works *much* better with the native Android scheduler. If you build `benchmark_model` from head, and build against arm64 (`--config=android_arm64`), you should see this when running benchmarks both from the command-line as well as in your app (particularly when using multi-threading). We're preparing a blog post about this new backend.

**Comment by arashb** [comment]
> One thing to note is that is that our legacy, multi-threaded floating point kernels are notoriously sensitive to unpinned execution, and generally work poorly with Android's thread scheduler.
> 
> We've been working on a [new gemm backend](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/BUILD#L193) which largely resolves this, and works _much_ better with the native Android scheduler. If you build `benchmark_model` from head, and build against arm64 (`--config=android_arm64`), you should see this when running benchmarks both from the command-line as well as in your app (particularly when using multi-threading). We're preparing a blog post about this new backend.

@jdduke I appreciate it if you share the link to the blog post.

**Comment by jdduke** [comment]
Hey @arashb, the blog post itself has been in flux, though we've highlighted the work a bit in recent talks, e.g., at [TFWorld](https://youtu.be/0SpZy7iouFU?t=627).  We're planning a more comprehensive performance blog post soon.

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this issue, as you are using an older version of tensorflow(1.x) which is officially considered as end of life. We recommend that you upgrade to 2.4 or later version and let us know if the issue still persists in newer versions. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, Please open a new issue for any help you need against 2.x, and we will get you the right help.
